glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,225 @@
+        let data_path = PathBuf::from_str(&data_path).expect("data_path must be a valid path.");
+        let local_queue = queue.clone();
+        let local_flag = processed_pending_pings.clone();
+        let _ = thread::Builder::new()
+            .name("glean.upload_manager.process_pings_directory".to_string())
+            .spawn(move || match process_pings_dir(&data_path) {
+                Ok(requests) => {
+                    let mut local_queue = local_queue
+                        .write()
+                        .expect("Can't write to pending pings queue.");
+                    local_queue.extend(requests.into_iter());
+                    local_flag.store(true, Ordering::SeqCst);
+                }
+                Err(e) => log::info!("Error processing pending pings directories! {}", e),
+            });
+
+        Self {
+            queue,
+            processed_pending_pings,
+        }
+    }
+
+    fn has_processed_pending_pings(&self) -> bool {
+        self.processed_pending_pings.load(Ordering::SeqCst)
+    }
+
+    /// Creates a `PingRequest` and adds it to the queue.
+    pub fn enqueue_ping(&self, uuid: &str, url: &str, body: JsonValue) {
+        let mut queue = self
+            .queue
+            .write()
+            .expect("Can't write to pending pings queue.");
+        let request = PingRequest::new(uuid, url, body);
+        queue.push_back(request);
+    }
+
+    /// Clear pending pings queue.
+    pub fn clear_ping_queue(&self) {
+        let mut queue = self
+            .queue
+            .write()
+            .expect("Can't write to pending pings queue.");
+        queue.clear();
+    }
+
+    /// Get the next `PingRequest` in queue.
+    pub fn get_next_ping(&self) -> Option<PingRequest> {
+        let mut queue = self
+            .queue
+            .write()
+            .expect("Can't write to pending pings queue.");
+        queue.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::metrics::PingType;
+    use crate::tests::new_glean;
+
+    #[test]
+    fn test_doesnt_error_when_there_are_no_pending_pings() {
+        let (glean, _) = new_glean(None);
+
+        // Try and get the next request
+        let request = glean.upload_manager.get_next_ping();
+
+        // Verify request was not returned
+        assert!(request.is_none());
+    }
+
+    #[test]
+    fn test_returns_ping_request_when_there_is_one() {
+        let (mut glean, _) = new_glean(None);
+
+        // Register a ping for testing
+        let ping_type = PingType::new("test", true, true);
+        glean.register_ping_type(&ping_type);
+
+        // Submit the ping to populate the pending_pings directory
+        glean.submit_ping(&ping_type).unwrap();
+
+        // Try and get the next request
+        let request = glean.upload_manager.get_next_ping();
+
+        // Verify request was returned
+        assert!(request.is_some());
+    }
+
+    // #[test]
+    // fn test_pings_not_sent_pings_are_loaded_on_init() {
```

Comment:
I decided to delete this test. I think there is really no reason to leave it there, since it doesn't work and will be different when I actually make it work. Is that alright?

Reply:
It is!